spring-web/src/main/java/org/springframework/http/converter/json/KotlinSerializationJsonHttpMessageConverter.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2002-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.http.converter.json;
+
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+
+import kotlinx.serialization.json.Json;
+import kotlinx.serialization.json.JsonConfiguration;
+import kotlinx.serialization.json.JsonDecodingException;
+import kotlinx.serialization.modules.EmptyModule;
+import kotlinx.serialization.modules.SerialModule;
+import org.jetbrains.annotations.NotNull;
+
+import org.springframework.http.HttpInputMessage;
+import org.springframework.http.HttpOutputMessage;
+import org.springframework.http.MediaType;
+import org.springframework.http.converter.AbstractHttpMessageConverter;
+import org.springframework.http.converter.HttpMessageNotReadableException;
+import org.springframework.http.converter.HttpMessageNotWritableException;
+import org.springframework.http.converter.KotlinSerializationResolver;
+import org.springframework.lang.Nullable;
+import org.springframework.util.StreamUtils;
+
+/**
+ * Implementation of {@link org.springframework.http.converter.HttpMessageConverter} that can read and write JSON using
+ * <a href="https://github.com/Kotlin/kotlinx.serialization">kotlinx.serialization</a>.
+ *
+ * <p>This converter can be used to bind {@code @Serializable} Kotlin classes. It supports {@code application/json} and
+ * {@code application/*+json} with various character sets, {@code UTF-8} being the default.
+ *
+ * @author Andreas Ahlenstorf
+ */
+public class KotlinSerializationJsonHttpMessageConverter extends AbstractHttpMessageConverter<Object> {
+
+	private static final Charset DEFAULT_CHARSET = StandardCharsets.UTF_8;
+
+	private final KotlinSerializationResolver resolver = new KotlinSerializationResolver();
+
+	private final Json json;
+
+	/**
+	 * Construct a new {@code KotlinSerializationJsonHttpMessageConverter} with default configuration.
+	 */
+	public KotlinSerializationJsonHttpMessageConverter() {
+		this(JsonConfiguration.getDefault(), EmptyModule.INSTANCE);
+	}
+
+	/**
+	 * Construct a new {@code KotlinSerializationJsonHttpMessageConverter} with custom configuration.
+	 */
+	public KotlinSerializationJsonHttpMessageConverter(JsonConfiguration jsonConfiguration, SerialModule serialModule) {
```

Comment:
Can we also get a constructor that just takes the ready build `Json` class? That way we're much more flexible. I'd even go as far as removing this constructor in favor of just `Json`.

Reply:
> That way we're much more flexible. I'd even go as far as removing this constructor in favor of just Json.

Could you please elaborate? Removing parameters/constructors is usually the opposite of "more flexible"?

Reply:
I replied but somehow it wasn't picked up correctly. It's here: https://github.com/spring-projects/spring-framework/pull/24436#discussion_r432472587